### ISSUE TITLE
diyone7 produces unreachable output states

### ISSUE THREAD

**Issue by chacon01** [herd/herdtools7#258]
The command
```
% diyone7 -ua 1 -arch AArch64 "PodWRPA Amo.LdAddAL RfeLP Amo.StClrPL DMB.SYdWWLP Wse"
```

Produces a litmus test with an unreachable output state:
```
AArch64 A
"PodWRPA Amo.LdAddAL RfeLP Amo.StClrPL DMB.SYdWWLP Wse"
Generator=diyone7 (version 7.55)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Rf Ws
Orig=PodWRPA Amo.LdAddAL RfeLP Amo.StClrPL DMB.SYdWWLP Wse
{
0:X1=x; 0:X2=y;
1:X0=y; 1:X3=x;
}
 P0                 | P1             ;
 MOV W0,#2          | MOV W1,#2      ;
 STR W0,[X1]        | STCLRL W1,[X0] ;
 MOV W4,#1          | DMB SY         ;
 LDADDAL W4,W3,[X2] | MOV W2,#1      ;
                    | STR W2,[X3]    ;
exists
(x=2 /\ y=2 /\ 0:X3=0)
```

Here, `y` is modified by the `LDADDAL` in P0 and the `STCLRL` in P1. Regardless of the order of the `LDADDAL` and `STCLRL`, the resulting value of `y` is always 1:
* `y` starts as 0, is incremented to 1 by the `LDADDAL`, and is unchanged by the `STRCLRL`
* `y` starts as 0, is unchanged by the `STRCLRL`, and is incremented to 1 by the `LDADDAL`

I've only observed this behavior with the `LDADD`, `LDEOR`, `LDSET`, and `LDCLR` (including store variants e.g. `STADD`).

---

Some additional examples:
```
% diyone7 -ua 1 -arch AArch64 "PodRRAA Amo.StAdd RfiPA HatAA Amo.LdAddAL PodWWLP RfePA"
```
which produces
```
AArch64 A
"PodRRAA Amo.StAddAP RfiPA HatAA Amo.LdAddAL PodWWLP RfePA"
Generator=diyone7 (version 7.55)
Prefetch=0:x=F,0:y=T,1:y=F,1:x=W
Com=Ws Rf
Orig=PodRRAA Amo.StAddAP RfiPA HatAA Amo.LdAddAL PodWWLP RfePA
{
0:X1=x; 0:X2=y;
1:X0=y; 1:X4=x;
}
 P0            | P1                 ;
 LDAR W0,[X1]  | MOV W2,#2          ;
 MOV W3,#1     | LDADDAL W2,W1,[X0] ;
 STADD W3,[X2] | MOV W3,#1          ;
 LDAR W4,[X2]  | STR W3,[X4]        ;
exists
(y=2 /\ 0:X0=1 /\ 0:X4=1 /\ 1:X1=1)
```
`y` must be 3 because P0 atomically increments it by 1 and P1 atomically increments it by 2.

---

```
% diyone7 -ua 1 -arch AArch64 "DpAddrdW Rfi Hat Amo.StEorPL DMB.SYdWWLP Rfe"
```
which produces
```
AArch64 A
"DpAddrdW Rfi Hat Amo.StEorPL DMB.SYdWWLP Rfe"
Generator=diyone7 (version 7.55)
Prefetch=0:x=F,0:y=T,1:y=F,1:x=W
Com=Ws Rf
Orig=DpAddrdW Rfi Hat Amo.StEorPL DMB.SYdWWLP Rfe
{
0:X1=x; 0:X4=y;
1:X0=y; 1:X3=x;
}
 P0                  | P1             ;
 LDR W0,[X1]         | MOV W1,#2      ;
 EOR W2,W0,W0        | STEORL W1,[X0] ;
 MOV W3,#1           | DMB SY         ;
 STR W3,[X4,W2,SXTW] | MOV W2,#1      ;
 LDR W5,[X4]         | STR W2,[X3]    ;
exists
(y=2 /\ 0:X0=1 /\ 0:X5=1)
```
`y` must be either 1 (P0's `STR` occurs after P1's `STEORL`) or 3 (P1's `STEORL` occurs after P0's `STR`).

**Comment by maranget** [comment]
Thanks for reporting the problem. We are aware of the problem, of which solution is not straightforward. However we can suggest a workaround.

Generators of the diy tools have very little knowledge of instruction semantics. They basically consider all writing instructions as plain stores writing values 1,2, 3 etc. This is of course insufficient for handling read-modify-write instructions such as `LDADD` and others. However, one may combine `herd7` and other tools to work around the problem. Let us consider your first example.

Let us first produce the test in file `X.litmus`
```
diyone7 -arch AArch64 -name X "PodWRPA Amo.LdAddAL RfeLP Amo.StClrPL DMB.SYdWWLP Wse"
```
We intend to produce a test whose condition will flag behaviours that are not sequentially consistent. As it is not easy to find those behaviours automatically, we'd rather produce a condition that checks SC compliance. To that aim, we run the test on top of the SC model and change the SC output into the appropriate condition:
```
% herd7 -cat sc.cat X.litmus > SC
% mlog2cond7 -optcond true -forall true SC > p.txt
% cat p.txt
X  "forall 0:X3=0 /\ y=1 /\ (x=1 \/ x=2)"
```
And finally we change the condition of the test to the new one:
```
% mkdir OUT
% recond7 -o OUT -conds p.txt X.litmus
% cat OUT/X.litmus
AArch64 X
"PodWRPA Amo.LdAddAL RfeLP Amo.StClrPL DMB.SYdWWLP Coe"
Generator=diyone7 (version 7.56+02~dev)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Rf Co
Orig=PodWRPA Amo.LdAddAL RfeLP Amo.StClrPL DMB.SYdWWLP Coe
{
0:X1=x; 0:X2=y;
1:X0=y; 1:X3=x;haba
}
 P0                 | P1             ;
 MOV W0,#2          | MOV W1,#2      ;
 STR W0,[X1]        | STCLRL W1,[X0] ;
 MOV W4,#1          | DMB SY         ;
 LDADDAL W4,W3,[X2] | MOV W2,#1      ;
                    | STR W2,[X3]    ;
forall 0:X3=0 /\ y=1 /\ (x=1 \/ x=2)
``` 
The test can now be run on hardware with `litmus7` or simulated on top of the standard aarch64 model. Any non-SC outcome will be flagged as a _negative_ witness.

**Comment by chacon01** [comment]
Thanks for the workaround. Unfortunately, I don't think it helps for my intended usage. The original test was supposed to be "forbidden" in that it asserted that a particular outcome was impossible subject to the constraints of the memory model. In the updated `forall` version, all orderings of events (including those forbidden by the memory model) will result in positive witnesses.

Since it's possible to detect unreachable outcomes automatically (e.g. by running with `herd7` on a memory model that has no ordering for acquire/release/dmb), I've started filtering out generated litmus tests that have them. That seems to be an effective workaround, but about a third of the generated litmus tests using atomics end up getting filtered out.

**Comment by maranget** [comment]
It looks a good idea to change the model from SC to anything that would fit. I can think of another adaptation of the technique: produce conditions that flag the difference between two models. Assume two models A.cat and B.cat, with B.cat being strictly weaker. That is all behaviours allowed by A.cat are allowed by B.cat and there exists behaviours that are allowed by B.cat and forbidden by A.cat. Let T.litmus be a test, first run T.litmus on top of the two models:
```
% herd7 -cat A.cat T.litmus > A
% herd7 -cat B.cat T.litmus > B
```
Then compare A and B saving the "positive" outcomes as a condition  in file p.txt:
```
% mcompare7 A B -cpos p.txt
```
Change the condition of T.litmus to the new condition:
```
% mkdir OUT
% recond7 -conds p.txt -o OUT T.litmus
```
Caveat: If you are operating on several files at a time (replacing T.litmus by an index file `@all` that lists the file names of the tests). There will be tests whose output in A and B will be the same. You probably want to eliminate those tests. The command sequence is then as follows:
```
% herd7 -cat A.cat @all > A
% herd7 -cat B.cat @all > B
% mcompare7 -pos P -cpos p.txt A B
% mkdir OUT
% recond7 -names P -conds p.txt -o OUT @jalglave
```
Where file `P` contains the names of the tests that effectively have more behaviours on top of model B.cat.

**Comment by hugookeeffe** [comment]
Pull request #294  has implemented the semantics for these atomic instructions into the generation for litmus test output states. I have tested the code with the examples given in the issue and it seems to be providing the correct desired output states now.
 
One thing to note, for the following command that was mentioned in the issue:
`% diyone7 -ua 1 -arch AArch64 "PodRRAA Amo.StAdd RfiPA HatAA Amo.LdAddAL PodWWLP RfePA"`
 
The first edge in the cycle (PodRRAA) should be PodRRAP. Using PodRRAA forces Amo.StAdd to have acquire semantics, which should not be allowed. In previous versions of the tools this would still work but it has since been fixed. The command should be:
`% diyone7 -ua 1 -arch AArch64 "PodRRAP Amo.StAdd RfiPA HatAA Amo.LdAddAL PodWWLP RfePA"`
 
If anything is unclear about my explanation, I would be happy to elaborate.
Hugo
